app/src/main/java/com/duckduckgo/app/browser/favicon/FaviconManager.kt

```
@@ -0,0 +1,186 @@
+        val domain = extractDomain(url)
+        val cachedFavicon = faviconPersister.faviconFile(FAVICON_TEMP_DIR, subFolder, domain)
+        if (cachedFavicon != null) {
+            faviconPersister.copyToDirectory(cachedFavicon, FAVICON_PERSISTED_DIR, NO_SUBFOLDER, domain)
+        }
+    }
+
+    override suspend fun deletePersistedFavicon(url: String) {
+        val domain = extractDomain(url)
+        val remainingFavicons = persistedFaviconsForDomain(domain) - 1 // -1 because we are about to delete one
+        if (remainingFavicons <= 0) {
+            faviconPersister.deletePersistedFavicon(domain)
+        }
+    }
+
+    override suspend fun deleteOldTempFavicon(subFolder: String, path: String?) {
+        faviconPersister.deleteFaviconsForSubfolder(FAVICON_TEMP_DIR, subFolder, path)
+    }
+
+    override suspend fun deleteAllTemp() {
+        faviconPersister.deleteAll(FAVICON_TEMP_DIR)
+    }
+
+    private suspend fun loadToViewFromDirectory(url: String, directory: String, subFolder: String, view: ImageView) {
+        val domain = extractDomain(url)
+        val cachedFavicon = getFaviconForDomainOrFallback(directory, subFolder, domain)
+        cachedFavicon?.let {
+            loadFaviconToView(cachedFavicon, view)
+        }
+    }
+
+    private suspend fun downloadFromUrl(url: String): Bitmap? {
+        return try {
+            val faviconUrl = getFaviconUrl(url)
+            faviconDownloader.getFaviconFromUrl(faviconUrl)
+        } catch (e: Exception) {
+            Timber.d(e, "Error downloading favicon")
+            null
+        }
+    }
+
+    private fun getFaviconUrl(url: String): Uri {
+        return if (url.toUri().host.isNullOrBlank()) {
+            "https://$url".toUri().faviconLocation() ?: throw IllegalArgumentException("Invalid favicon currentPageUrl")
+        } else {
+            url.toUri().faviconLocation() ?: throw IllegalArgumentException("Invalid favicon currentPageUrl")
+        }
+    }
```

two questions on this function

1. what is `currentPageUrl` - should that be a variable instead of a string literal?
2. do we need to throw the exception? could we make the function return nullable type `Uri?` and return `null` if we can't get extract a favicon URL from the url string? 

I don't feel too strongly about it, but the exception feels a little heavy handed, and given the exceptions are runtime and not checked, i think the `nullable` return type is safer here too, as we're forced to handle that and can't forget to catch the exception

---

Both things are coming from the old favicon code where we threw that same exception. It is captured in `downloadFromUrl` which is capturing at the same time exceptions from Glide. I've changed it to return null and wrapped the glide call with `runCatching`